File: pulsar-ql-server/src/main/java/fun/platonic/pulsar/ql/SessionConfig.java
package fun.platonic.pulsar.ql;

import fun.platonic.pulsar.common.config.ImmutableConfig;
import fun.platonic.pulsar.common.config.VolatileConfig;

public class SessionConfig extends VolatileConfig {

    private DbSession dbSession;

    public SessionConfig(DbSession dbSession, ImmutableConfig fallbackConfig) {
        super(fallbackConfig);
        this.dbSession = dbSession;
    }

    @Override
    public void setTTL(String name, int ttl) {
        if (ttl > 0) {
            ttl = 1 + ttl + dbSession.getSqlSequence();
        }

        super.setTTL(name, ttl);
    }

    @Override
    public boolean isExpired(String propertyName) {
        int sequence = dbSession.getSqlSequence();
        int ttl = getTTL(propertyName);
        boolean expired = sequence > ttl;
        // LOG.debug("Property {}, sequence: {}, ttl: {}", propertyName, sequence, ttl);
        if (LOG.isDebugEnabled() && expired) {
            LOG.debug("Property {} is expired at the {}th command", propertyName, sequence);
        }

        return expired;
    }

}

File: pulsar-ql-server/src/main/kotlin/fun/platonic/pulsar/ql/SessionConfig.kt
package `fun`.platonic.pulsar.ql

import `fun`.platonic.pulsar.common.config.ImmutableConfig
import `fun`.platonic.pulsar.common.config.VolatileConfig

class SessionConfig(private val dbSession: DbSession, fallbackConfig: ImmutableConfig): VolatileConfig(fallbackConfig) {

    override fun setTTL(name: String, ttl: Int) {
        super.setTTL(name, 1 + ttl + dbSession.sqlSequence)
    }

    override fun isExpired(propertyName: String): Boolean {
        val sequence = dbSession.sqlSequence
        val ttl = super.getTTL(propertyName)
        val expired = sequence > ttl
        // log.debug("Property {}, sequence: {}, ttl: {}", propertyName, sequence, ttl);
        if (LOG.isDebugEnabled && expired) {
            LOG.debug("Property {} is expired at the {}th command", propertyName, sequence)
        }

        return expired
    }
}
